contracts/GenesisManager.sol

```diff
@@ -0,0 +1,268 @@
+    function addTeamGrant(
+        address _receiver,
+        uint256 _amount,
+        uint256 _timeToCliff,
+        uint256 _vestingDuration
+    )
+        external
+        onlyOwner
+        atStage(Stages.GenesisStart)
+    {
+        uint256 updatedGrantsAmount = teamGrantsAmount.add(_amount);
+        // Amount of tokens included in team grants cannot exceed the team supply during genesis
+        require(updatedGrantsAmount <= teamSupply);
+
+        teamGrantsAmount = updatedGrantsAmount;
+
+        addVestingGrant(_receiver, _amount, _timeToCliff, _vestingDuration);
+    }
+
+    /**
+     * @dev Add an investor grant for tokens with a vesting schedule
+     * @param _receiver Grant receiver
+     * @param _amount Amount of tokens included in the grant
+     * @param _timeToCliff Seconds until the vesting cliff
+     * @param _vestingDuration Seconds starting from the vesting cliff until the end of the vesting schedule
+     */
+    function addInvestorGrant(
+        address _receiver,
+        uint256 _amount,
+        uint256 _timeToCliff,
+        uint256 _vestingDuration
+    )
+        external
+        onlyOwner
+        atStage(Stages.GenesisStart)
+    {
+        uint256 updatedGrantsAmount = investorsGrantsAmount.add(_amount);
+        // Amount of tokens included in investor grants cannot exceed the investor supply during genesis
+        require(updatedGrantsAmount <= investorsSupply);
+
+        investorsGrantsAmount = updatedGrantsAmount;
+
+        addVestingGrant(_receiver, _amount, _timeToCliff, _vestingDuration);
+    }
+
+    /**
+     * @dev Add a grant for tokens with a vesting schedule. An internal helper function used by addTeamGrant and addInvestorGrant
+     * @param _receiver Grant receiver
+     * @param _amount Amount of tokens included in the grant
+     * @param _timeToCliff Seconds until the vesting cliff
+     * @param _vestingDuration Seconds starting from the vesting cliff until the end of the vesting schedule
+     */
+    function addVestingGrant(
+        address _receiver,
+        uint256 _amount,
+        uint256 _timeToCliff,
+        uint256 _vestingDuration
+    )
+        internal
+    {
+        // Receiver must not have already received a grant with a vesting schedule
+        require(vestingHolders[_receiver] == address(0));
+
+        // The grant's vesting schedule starts when the token distribution ends
+        uint256 startTime = tokenDistribution.getEndTime();
+        // Create a vesting holder contract to act as the holder of the grant's tokens
+        // Note: the vesting grant is revokable
+        TokenVesting holder = new TokenVesting(_receiver, startTime, _timeToCliff, _vestingDuration, true);
+        vestingHolders[_receiver] = holder;
+
+        // Transfer ownership of the vesting holder to the owner of this contract
+        // giving the owner of this contract to revoke the grant
+        holder.transferOwnership(owner);
+
+        token.transfer(holder, _amount);
```

Comment:
look at TokenVesting contract to see how it is aware of how much token is initially granted, and therefore how it knows how much to vest. What if more token is sent to the vesting contract after its begun.